### ISSUE TITLE
Errors appears while using MinGW to build OMR on Windows

### ISSUE THREAD

**Issue by samolisov** [eclipse/omr#2370]
I try using *MinGW-w64* as a C/C++ toolchain on Windows to build *OMR* using `CMake` and a number of errors appear, for example:

```bash
> cmake --build .

C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/7.3.0/../../../../x86_64-w64-mingw32/lib/../lib/libmingw32.a(lib64_libmingw32_a-cr
t0_c.o): In function `main':
C:/repo/mingw-w64-crt-git/src/mingw-w64/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [tools/hookgen/CMakeFiles/hookgen.dir/build.make:123: tools/hookgen/hookgen.exe] Error 1
make[1]: *** [CMakeFiles/Makefile2:273: tools/hookgen/CMakeFiles/hookgen.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

I've found the solution for this problem: the linker flag `-municode` must be added to the `OMR_PLATFORM_EXE_LINKER_OPTIONS` list to enable the support of the Unicode version of `WinMain`. OK, I'm able to do this. But then I get the following, more complicated, problem:

```bash
> cmake --build .

[  5%] Building C object thread/CMakeFiles/j9thr_obj.dir/common/thrprof.c.obj
C:/openj9/omr/thread/common/thrprof.c: In function 'paint_stack':
C:/openj9/omr/thread/common/thrprof.c:558:2: error: '__try' undeclared (first use in this function); did you mean '__try1'
?
  __try {
  ^~~~~
  __try1
C:/openj9/omr/thread/common/thrprof.c:558:2: note: each undeclared identifier is reported only once for each function it a
ppears in
C:/openj9/omr/thread/common/thrprof.c:558:8: error: expected ';' before '{' token
  __try {
        ^
make[2]: *** [thread/CMakeFiles/j9thr_obj.dir/build.make:386: thread/CMakeFiles/j9thr_obj.dir/common/thrprof.c.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:498: thread/CMakeFiles/j9thr_obj.dir/all] Error 2
make: *** [Makefile:141: all] Error 2
```

I've looked for a solution carefully and found the following answers on SO: [this](https://stackoverflow.com/questions/18101801/why-doesnt-gcc-compile-this-code) and [this](https://stackoverflow.com/questions/12513209/try-and-exception-portability). As I understand, `__try/__finally` is not a portable construction for C. Of course, we can just define these keywords as the Ruby guys did it (please see commit [6e066e](https://github.com/chef/win32-service/commit/6e066e3845f745ede9e8c77f1250bc30e2e5d85a) - they check whether `seh.h` is defined) 
 or just include the `excpt.h` header (checking the availability of the file using the `CHECK_INCLUDE_FILES` macros from the `CheckIncludeFiles` CMake library) but I'm **not sure** if the above options will work. 

Any comments are welcome - *MinGW* is smaller than *Microsoft Visual C++ Build Tools* and the idea to use this toolchain - *MinGW* and *Eclipse CDT* - for the building process on Windows seems interesting.

**Comment by amitin** [comment]
@samolisov __try/__except/__finally is a Microsoft extension to C for supporting Windows Structured Exception Handling (SEH). MinGW doesn't support it for a long time. Some workarounds exist but none of them is suitable to use in production-quality software. 
That's why it's better to use Clang for Windows - it's has full support of SEH and the example above compiles and works just fine.

**Comment by samolisov** [comment]
@amitin Thank you very much for your answer. I'm a bit confused: I've spent some time trying to build OMR using CLang for Windows ('clang-cl') and [Ninja](https://ninja-build.org/) only but it looks like CLang is absolutely not suitable without [Visual C++ Build Tools](http://landinghub.visualstudio.com/visual-cpp-build-tools) (these tools take about 5 GB HDD just to be installed!) because the libs needed for linking: kernel32, user32, etc. are there. Visual C++ Build Tools contain its own compilator and linker and, as I see, OMR has the `cmake\modules\platform\toolcfg\msvc.cmake` file that configures Visual C++ Toolchain. So the question is: does it matter to use CLang vs MSVC and which one is taken into account for OpenJ9 (and OMR respectively) nightly builds?

**Comment by amitin** [comment]
@samolisov, in our project we use clang as a custom build tool for MSVC, this works fine because clang-cl mimics the standard MS cl.exe compiler. And I think you would need Windows standard libs linked any way, no matter which compiler you use.
For your first question it's up to you: if you don't need something which standard cl.exe doesn't support (like 'computed goto') then I guess it's better to use the standard cl.exe.
Sorry, your second question is beyond my expertise, I'm not envovled into any build process for OMR/J9.

**Comment by samolisov** [comment]
@amitin Thank again for your help. I've downloaded and installed `MSVC` to my environment and can confirm `OMR` is able to be built using this toolchain. `Clang` is a bit more rigorous, so I get a *comment in comment* error (a PR #2373 is open, I believe it solves this). 

In order to make the life a little bit easy for Windows users who would like to build `OMR` on their environments, I've added some instruction to the `BuildingWithCMake.md` file. You opinion why *MinGW* shouldn't be used to build `OMR` is included.

This issue will be closed when maintainers merge #2374.
